test/Microsoft.AspNet.Mvc.Core.Test/SessionStateTempDataProviderTest.cs

```diff
@@ -201,20 +189,150 @@ public void SaveAndLoad_SimpleTypesCanBeStoredAndLoaded()
+            {
+                { "DateTime", inputDatetime }
+            };
+            var context = GetHttpContext(new TestSessionCollection(), true);
+
+            // Act
+            testProvider.SaveTempData(context, input);
+            var TempData = testProvider.LoadTempData(context);
+
+            // Assert
+            var datetime = Assert.IsType<DateTime>(TempData["DateTime"]);
+            Assert.Equal(inputDatetime, datetime);
+        }
+
+        [Fact]
+        public void SaveAndLoad_GuidCanBeStoredAndLoaded()
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+            var inputGuid = Guid.NewGuid();
+            var input = new Dictionary<string, object>
+            {
+                { "Guid", inputGuid }
+            };
+            var context = GetHttpContext(new TestSessionCollection(), true);
+
+            // Act
+            testProvider.SaveTempData(context, input);
+            var TempData = testProvider.LoadTempData(context);
+
+            // Assert
             var guidVal = Assert.IsType<Guid>(TempData["Guid"]);
-            Assert.Equal(inputGuid.ToString(), guidVal.ToString());
+            Assert.Equal(inputGuid, guidVal);
+        }
+
+        [Fact]
+        public void SaveAndLoad_ListCanBeStoredAndLoaded()
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+            var input = new Dictionary<string, object>
+            {
+                { "List`string", new List<string> { "one", "two" } }
+            };
+            var context = GetHttpContext(new TestSessionCollection(), true);
+
+            // Act
+            testProvider.SaveTempData(context, input);
+            var TempData = testProvider.LoadTempData(context);
+
+            // Assert
             var list = (IList<string>)TempData["List`string"];
             Assert.Equal(2, list.Count);
-            Assert.Equal("one", list[0]);
-            Assert.Equal("two", list[1]);
+        }
+
+        [Fact]
+        public void SaveAndLoad_DictionaryCanBeStoredAndLoaded()
+        {
+            // Arrange
+            var testProvider = new SessionStateTempDataProvider();
+            var inputDictionary = new Dictionary<string, string>
+            {
+                { "Hello", "World" },
+            };
+            var input = new Dictionary<string, object>
+            {
+                { "Dictionary", inputDictionary }
+            };
+            var context = GetHttpContext(new TestSessionCollection(), true);
+
+            // Act
+            testProvider.SaveTempData(context, input);
+            var TempData = testProvider.LoadTempData(context);
+
+            // Assert
             var dictionary = Assert.IsType<Dictionary<string, string>>(TempData["Dictionary"]);
             Assert.Equal("World", dictionary["Hello"]);
+        }
+
+        [Fact]
+        public void SaveAndLoad_EmptyDictionaryCanBeStoredAndLoaded()
```

Comment:
change name to `SaveAndLoad_EmptyDictionary_RoundTripsAsNull()` since the original value (the empty dictionary) is _not_ loaded.